### PR TITLE
fix: generation of memoized class member calls

### PR DESCRIPTION
**Pull request by WinPlay02**
- fix: generating memoized class member calls (on non-static members) should still use the class for python code generation
- added testcase
------------------
Fixes the (last) issue in #954

### CODE REVIEW

File: packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts [hunk 1]
```diff
@@ -1083,6 +1090,10 @@ export class SafeDsPythonGenerator {
         throw new Error('Callable of provided call does not exist or is not a declaration.');
     }
 
+    private getClassQualifiedNameForMember(expression: SdsClassMember): string {
+        return `${this.getPythonNameOrDefault(AstUtils.getContainerOfType(expression.$container, isSdsDeclaration)!)}.${this.getPythonNameOrDefault(expression)}`;
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts:1094]: Classes can be nested, which is not handled here:

```
class Outer {
    class Nested {
        @Pure
        static fun f() -> result: Boolean
    }
}
```

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts (hunk 1 above)
Comment (WinPlay02) [reply packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts]: If understood correctly, this should now be fixed

File: packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts [hunk 2]
```diff
@@ -1083,6 +1090,16 @@ export class SafeDsPythonGenerator {
         throw new Error('Callable of provided call does not exist or is not a declaration.');
     }
 
+    private getClassQualifiedNameForMember(expression: SdsClassMember): string {
+        const classMemberPath = [this.getPythonNameOrDefault(expression)];
+        let enclosingClass = AstUtils.getContainerOfType(expression.$container, isSdsClass);
+        while (enclosingClass) {
+            classMemberPath.push(this.getPythonNameOrDefault(enclosingClass));
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts:1097]: Could also `unshift` here to insert at the start. Then you don't need to reverse at the end.

File: packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts [hunk 3]
```diff
@@ -1083,6 +1090,16 @@ export class SafeDsPythonGenerator {
         throw new Error('Callable of provided call does not exist or is not a declaration.');
     }
 
+    private getClassQualifiedNameForMember(expression: SdsClassMember): string {
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts:1093]:
```suggestion
    private getClassQualifiedNameForMember(node: SdsClassMember): string {
```

File: packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts [hunk 4]
```diff
@@ -1083,6 +1090,16 @@ export class SafeDsPythonGenerator {
...
+        const classMemberPath = [this.getPythonNameOrDefault(expression)];
+        let enclosingClass = AstUtils.getContainerOfType(expression.$container, isSdsClass);
+        while (enclosingClass) {
```
Comment (lars-reimann) [inline packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts]: Last refactoring:

```suggestion
        const classMemberPath = [];
        let current = node;
        while (enclosingClass) {
```

Then the logic to update `current` is only needed once.

**Review by lars-reimann — APPROVED** [review]
Excellent, thank you.

**Pull request merged**